tools/adaparse.cpp

```diff
@@ -1,8 +1,147 @@
 #include <iostream>
 #include <string_view>
+#include <sys/stat.h>
+#include <fmt/core.h>
 #include <cxxopts.hpp>
+#include <fstream>
 #include "ada.h"
+#include "line_iterator.h"
+#ifdef _MSC_VER
+#include <io.h>
+#else
+#include <unistd.h>
+#endif
+
+uint64_t nano() {
+  return std::chrono::duration_cast<::std::chrono::nanoseconds>(
+             std::chrono::steady_clock::now().time_since_epoch())
+      .count();
+}
+
+
+bool print_part(std::string_view get_part, const ada::url_aggregator& url) {
+    if(get_part.size() == 4) {
+        if(get_part[0] == 'h') {
+          if(get_part == "host") {
+            fmt::print("{}\n", url.get_host());
+            return true;
+          } else if(get_part == "hash") {
+            fmt::print("{}\n", url.get_hash());
+            return true;
+          }
+        } else if(get_part[0] == 'p') {
+          if(get_part == "port") {
+            fmt::print("{}\n", url.get_port());
+            return true;
+          }
+        }
+    } else if(get_part.size() == 6) {
+        if(get_part == "origin") {
+          fmt::print("{}\n", url.get_origin());
+          return true;
+        }
+        if(get_part == "search") {
+          fmt::print("{}\n", url.get_search());
+          return true;
+        }
+    } else if(get_part.size() == 7) {
+      if(get_part[0] == 'p') {
+        if(get_part == "protocol") {
+          fmt::print("{}\n", url.get_protocol());
+          return true;
+        }
+        if(get_part == "password") {
+          fmt::print("{}\n", url.get_password());
+          return true;
+        }
+        if(get_part == "pathname") {
+          fmt::print("{}\n", url.get_pathname());
+          return true;
+        }
+
+    }
+    } else if(get_part == "hostname") {
+        fmt::print("{}\n", url.get_hostname());
+        return true;
+
+    } else if(get_part == "username") {
+        fmt::print("{}\n", url.get_username());
+        return true;
+    }
+
+    fmt::print(stderr, "\"{}\" not found\n", get_part);
+    return false;
+}
+
+int piped_file(const cxxopts::ParseResult result, std::istream& input_stream) {
+  constexpr size_t cache_length = 32768;
+  std::unique_ptr<char[]> cachebuffer(new char[cache_length]{});
+
+  uint64_t before = nano();
+
+  size_t howmany = 0;
+  size_t tr;
+  size_t offset = 0;
+  size_t lines = 0;
+  size_t total_bytes = 0;
+  size_t blocks = 0;
+
+
+  std::string get_part; 
+  bool get_flag;
+  if (result.count("get")) {
+
+    get_part = result["get"].as<std::string>();
+
+  while ((tr = read(0, cachebuffer.get() + offset, cache_length - offset))) {
+    howmany += tr; 
+    blocks++; 
+    size_t capacity = tr + offset; 
+
+
+    line_iterator li(cachebuffer.get(), capacity); 
+    while (li.find_another_complete_line()) {
+      std::string_view line = li.grab_line(); 
+
+      ada::result<ada::url_aggregator> url = ada::parse(line);
+      if (!url) {
+        printf("Invalid URL: %.*s\n", int(line.size()), line.data());         
```

Comment:
Let us not mix `printf` and `fmt::` in the same code. Let us be modern and use `fmt::`.